### ISSUE TITLE
File Type Licenses

### ISSUE THREAD

**Issue by MarvinZacher** [sensslen/nuget-license#56]
When I add a package as a dependency which contains a commercial license I get the error `Validation for licenses of type File not yet supported`

## Dependency.csproj
```xml
...
<PropertyGroup>
<PackageLicenseFile>LICENSE.txt</PackageLicenseFile>
</PropertyGroup>
...
<ItemGroup>
<None Include="LICENSE.txt" Pack="true" PackagePath="" />
</ItemGroup>
...
```

## Expected result

In the output I'd expect the json output to be something like
```json
{
...
"License": "<first line of LICENSE.txt>",
"LicenseHeading": "<first line of LICENSE.txt, maybe this if someone parses the License field for valid license expressions>",
"LicenseUrl": "<empty or maybe from an --license-title-to-licenseurl-mappings map?>",
...
}
```
And the -d argument to copy the contents of the LICENSE.txt.

## Remarks
As Microsoft deprecated PackageLicenseUrl only File and Expression are left. But an expression doesn't work with a custom license and results in https://learn.microsoft.com/en-us/nuget/reference/errors-and-warnings/nu5124.

I've seen that there is a override argument but this won't help if I have lots and lots of dependencies. My usecase would be automated license checking&storage in CI Pipelines.

**Comment by sensslen** [comment]
Hi @MarvinZacher, the only option you have for now is to use the override method. Implementing a way to determine the license from a file is a challenging task that has not yet been done for nuget-license. Feel free to come up with s solution and provide a PR. I'll be happy to look at it. Note that the license text will not always match 100% though, as there are fields in almost every license to contain the author and such. All this needs to be accounted for. As far as I know, GitHub provides an API to determine the license of a project so it implements a way to determine the license. Maybe this would be a viable option.

The solution you propose I think would not be a good one, as nuget-license has always been about validating licenses against a set of allowed licenses. Therefore simply printing the license will not cut it, as there is no way to verify whether the license text matches an allowed license or not.

**Comment by kchenery** [comment]
This is a feature I am interested in too. Seems a lot of older packages use a license file approach rather than setting their license in the meta data.  That to me means you'd ahve to parse the file to infer the type from its contents. In some cases thats not too bad but in others I've seen mutiple licenses referenced and which one would you use.

Im interested in helping out here and am thinking about how it might be tackled.

If the file specifies the SPDX license type I think you can safely use that when reporting the license type of the nuget package. And then referencing that for the allowed/disallowed results.

**Comment by sensslen** [comment]
@kchenery I know github does show license information and they do so  reading the license file. Maybe there is an API that can do that? I have not yet invested the time to research how to do all this as the tool was good enough as it stands with occasional overwrites.

**Comment by kchenery** [comment]
Im worried about making web requests for a couple of reasons:

* hitting rate limits
* being behind some authentication/authorisation (eg private repo in the case of Github) or a prorpietary license behind a paywall from a company (which is something we face)
* being in some "air gapped" kind of environment where the request just fails (somewhat weird for adding a nuget package 😄)

An idea I've had is to be able to supply a list of license file paths (locally) to be able to validate the contents against. Still working through this idea in my head though to see if its plausible or just plain silly. It would mean having to have them to hand. But I dont think that is _too_ bad myself.

Our particular use case (our one 😉) for this tool is to run it in a Github action to check that developers havent inadvertently added a package that has a license we do not allow. In this case it would be pretty easy to handle those files as we'd just add them to the action repo.

**Comment by kchenery** [comment]
Any thoughts on the above @sensslen?

I'm giving it a go... I've got it loading the file contents to the package meta data object. So now can compare it. Im wondering if supplying a list of files to compare against is sensible or not.  I'd love to have it try and decipher the license from the content... but that might have to be some future work. AI could probably do it!

Thinking something like a json file to pass in:

```json
[
  {
    "Path": "./licenses/foo_proprietary_license.txt",
    "LicenseExpression": "LicenseRef-Foo"
  },
  {
    "Path": "./licenses/some_company_eula.txt",
    "LicenseExpression": "LicenseRef-SomeCompany-Eula"
  }
]
```

Tool could iterate over these entries, loading the files. Then these could then be added as accepted license types to the tool and report as such.

**Comment by sensslen** [comment]
Sure - sounds good to me - what has kept me from doing things like that was that most licenses contain some header that is specific to the project (like copyright information). Therefore comparing against a generic license most often does not cut it.... If you can work around that I'm more than happy to include this in the project....

**Comment by kchenery** [comment]
I'm unsure if I can... but its something I'd like too. I think a good first step is allowing people to grab those licenses and being able to state they're acceptable/allowed.

**Comment by sensslen** [comment]
> I'm unsure if I can... but its something I'd like too. I think a good first step is allowing people to grab those licenses and being able to state they're acceptable/allowed.

Sure - however this has questionable benefit, as it is pretty much equivalent to the override already avaialble....

**Comment by kchenery** [comment]
Thats fair... though we definitely have a use case for it 😄 

I decided to have a go at determining the license from the file contents itself. And I have had some success with it. With some basic testing I was able to get this:
```
+-----------------------+---------+----------------------------+----------------------+------------------------------------+---------------+
| Package               | Version | License Information Origin | License Expression   | License Url                        | Authors       |
+-----------------------+---------+----------------------------+----------------------+------------------------------------+---------------+
| Kent.NugetLicenceTest | 1.0.0   | PackageFile                | Unknown License Type | https://aka.ms/deprecateLicenseUrl | Kent Chenery  |
| NugetPackage2         | 1.2.3   | PackageFile                | MIT                  | https://aka.ms/deprecateLicenseUrl | NugetPackage2 |
+-----------------------+---------+----------------------------+----------------------+------------------------------------+---------------+
```

Its not doing exact matches, even though MIT is often unchanged. Its doing a bit of a best guess, with a threshold so there is potential for false positives but I think its acceptable.

To me, the big downsize with the way I have done this is that some license "templates" have to be embedded into the tool. So we could cover most common ones (like you are today with the package stating it in the expression).

**Comment by kchenery** [comment]
Im a little bit stuck with the tests. Im expecting some errors from existing tests but I get thousands! How are these files generated? Do I have to fix them all up?  How do I go about that - there must be some automated way.

**Comment by sensslen** [comment]
Once the tests have run, there will be two files - the one with the verified postfix and one I forget the postfix about. You can go ahead and fix the tests by just removing all verified files and renaming the other ones accordingly. I know I still have to work on making the generation not change even though the license information origin has changed.....

**Comment by kchenery** [comment]
OK... I think I have a bigger issue though. I rebased on your recent changes on `main` and thats broken things substantially for me. I'm developing on linux so .net 4.7.2 is a challenge. The new `NuGetLicense` project just wont load for me.

Is there a great need to support 4.7.2?

**Comment by kchenery** [comment]
Oh... a clean and new build seems to have fixed it.

**Comment by sensslen** [comment]
Yes there is a need for 4.7.2. Unfortunately projects containing native projects won't work properly in .net.....

**Comment by kchenery** [comment]
Have raised a PR (#240) that takes a stab at addressing this request.

**Comment by sensslen** [comment]
Hey guys, I know this was a long time and popular request. I'm happy to announce that https://github.com/sensslen/nuget-license/releases/tag/v4.0.0-beta.1 is the first version of nuget-license to support file type license matching (against the SPDX license identifiers). Currently there is no support for specifying custom license texts.
